torch/_subclasses/functional_tensor.py

```diff
@@ -537,6 +559,29 @@ def unwrap(x):
                         torch.Tensor, wrap, outs_unwrapped
                     )
                 else:
+                    arg_mutation_counters = []
+
+                    if self.export:
+                        # Check args before the operation
+                        for arg in args:
+                            if isinstance(
+                                arg, FunctionalTensor
+                            ) and torch._is_functional_tensor(arg.elem):
+                                arg_storage = StorageWeakRef(arg.untyped_storage())
+                                arg_group_id = self._partial_frozen_storage.get(
+                                    arg_storage, None
+                                )
+                                if (
+                                    arg_group_id in self._mutated_partial_frozen_storage
+                                ):  # type: ignore[attr-defined]
+                                    raise RuntimeError(
```

Comment:
hmm, how does this code ensure that you do **not** error on code like the below?
```
def f(x):
    y = x + 1
    z = y.to(...)
    z.mul_(2)
    return z
```

I was expected that once you mark a storage as "frozen but mutated", you also remember the specific tensor that was mutated (z in the above example), so you know that z itself is still safe to read from

Reply:
I see. I'll update the PR to error on this case. I didn't consider this case before, thanks for the catch!

Reply:
@bdhirsh If we allow this case, I guess we also allow the following case?

```
            def forward(self, x):
                y = x + 1
                z = y.to(torch.int32)
                z.mul_(2)
                z.add_(2)
                return z
```

Technically, `z.add_(2)` is using `z` in a computation. Can we say that we always allow using mutated storage in computation (or return) if the only tensor that mutates this storage is the tensor used in the computation itself?

Reply:
@bdhirsh I tried to implement this, but now I think it might not be possible in the new training IR.

Given this module:

```python
        class Module(torch.nn.Module):
            def forward(self, x):
                z = x + 1
                y = z.to("cpu")
                z.add_(1)
                return y
```

The training IR graph looks like this:

```
graph():
    %x : [num_users=1] = placeholder[target=x]
    %add : [num_users=1] = call_function[target=torch.ops.aten.add.Tensor](args = (%x, 1), kwargs = {})
    %to : [num_users=1] = call_function[target=torch.ops.aten.to.dtype_layout](args = (%add,), kwargs = {dtype: torch.int64, layout: torch.strided, device: cpu})
    %add_ : [num_users=1] = call_function[target=torch.ops.aten.add_.Tensor](args = (%to, 1), kwargs = {})
    return (add_,)
```

Specifically, originally we are doing z.add_(1), but the training IR graph has add_ on the result of to instead of the input. I think this is expected behavior. So we don't really know if we are returning `y` or `z`....

To keep the behavior of export consistent between all modes, maybe we just don't allow this case?

Reply:
@yushangdi oh the example you put above is a bit different: you are mutation `z`, but returning `y`. I agree that we need to ban that behavior (because it is semantically different to the user, depending on whether z and y are aliased).

In the example I gave higher up though, we mutate `z`, and also return `z`. This is ok because we never read from `y`, so it doesn't actually matter if y and z properly aliased or not.

Reply:
> @yushangdi oh the example you put above is a bit different: you are mutation `z`, but returning `y`. I agree that we need to ban that behavior (because it is semantically different to the user, depending on whether z and y are aliased).
> 
> In the example I gave higher up though, we mutate `z`, and also return `z`. This is ok because we never read from `y`, so it doesn't actually matter if y and z properly aliased or not.

@bdhirsh Yes, the module code is different, but if you look at the fx.graph generated, it's the same as the example you put above. My point is we cannot distinguish the two examples after they are converted to an fx.Graph by export_for_training.

Reply:
@tugsbayasgalan @bdhirsh 

What do you mean by `y is x`? Do you mean they are the same FunctionalTensor? 
If we do this, wouldn't we error on `z.mul_(2)` in both examples?

> (1) agree to ban both of the above examples (@tugsbayasgalan - my understanding is that some users hit the pessimistic error from example 1 above, so it would be preferable to not error on it?)

If we can't fix the examples above, I'm also wondering if it's worth to land this PR? With this PR, people can mutate the result of aten.to(), but they cannot use the result in any computation, seems like it's still quite restrictive and we had to add a bunch of code. If people are mostly hitting the errors in this example, that seems like landing this PR wouldn't benefit them...

Reply:
> What do you mean by y is x? Do you mean they are the same FunctionalTensor?

yep

> If we do this, wouldn't we error on z.mul_(2) in both examples?

and yep (although - we already error on both of those examples today, as you pointed out)

Reply:
> and yep (although - we already error on both of those examples today, as you pointed out)

I see. So you are saying, when `y` and `x` are the same FunctionalTensor, there's nothing we can do. However, if `y` and `x` are not the same FunctionalTensor, we can further relax it there and allow this case?

Reply:
yep!

Reply:
@bdhirsh do you have the example where user runs into this issue? I thought their issue was lot simpler haha... If the issue they ran into is indeed simpler, it seems ok to ban both.